# Forecasting Future Gains: Identifying Top 3 Investment Zipcodes in the Baltimore-Washington Area Using Zillow Time Series Data

- 07/18/24


### Introduction

In an era where data-driven decision-making is paramount, real estate investments stand to benefit significantly from advanced data science techniques. This project aims to leverage Zillow zipcode data to identify the top three zipcodes in the Baltimore-Washington area that promise the greatest return on investment (ROI) three years post-purchase. By employing the CRISP-DM (Cross Industry Standard Process for Data Mining) methodology, we will ensure a structured and thorough approach to our analysis, ensuring reliable and actionable insights for potential investors.



### CRISP-DM Overview

The CRISP-DM methodology consists of six major phases: Business Understanding, Data Understanding, Data Preparation, Modeling, Evaluation, and Deployment. Each phase is integral to the success of the project and is described in detail below.

#### 1. Business Understanding

**Objective:**  
The primary objective of this project is to identify the top three zipcodes in the Baltimore-Washington area that are predicted to offer the highest return on investment (ROI) for home buyers over a three-year period. The purchase cost is negligible due to the

**Key Questions:**
- What zipcodes have shown the greatest increase in typical home value, as assessed by Zillow, in the Baltimore-Washington area over the last 5 years ?
    - Of these zipcodes, select the top `5-10` that showed the highest percent increase in home value over the past 5 years.
    - `Consider the invesement cost of these zipcodes`
- What are the top three zipcodes that show the highest potential for ROI when forecasting 3 years into the future?

**Success Criteria:**
- Identification of top three zipcodes with the highest predicted ROI.
- Development of a robust predictive model with high accuracy and reliability.



#### 2. Data Understanding

**Objective:**  
To gather and understand the data required for the analysis. This involves exploring Zillow's zipcode-level data and other relevant datasets.

**Tasks:**
- Collect Zillow data, including historical home prices, sales trends, and economic indicators for the Baltimore-Washington area.
- Explore additional data sources such as crime rates, school quality, and local amenities.
- Conduct initial data exploration to understand the distribution, trends, and potential issues within the data.

**Deliverables:**
- Data collection report detailing sources, data types, and initial findings.
- Exploratory data analysis (EDA) report with visualizations and summary statistics.



#### 3. Data Preparation

**Objective:**  
To clean, preprocess, and transform the data for analysis.

**Tasks:**
- Handle missing values, outliers, and inconsistencies in the data.
- Normalize and standardize numerical features.
- Encode categorical variables.
- Feature engineering to create new variables that may enhance the predictive power of the model.

**Deliverables:**
- Cleaned and preprocessed dataset ready for modeling.
- Feature engineering documentation.



#### 4. Modeling

**Objective:**  
To develop predictive models that can accurately forecast the ROI for different zipcodes.

**Tasks:**
- Select appropriate modeling techniques (e.g., linear regression, decision trees, ensemble methods).
- Train and validate models using historical data.
- Optimize model parameters and assess performance using metrics such as RMSE, MAE, and R-squared.

**Deliverables:**
- Trained predictive models.
- Model performance report with evaluation metrics.



#### 5. Evaluation

**Objective:**  
To evaluate the performance of the models and ensure they meet the business objectives.

**Tasks:**
- Compare model predictions with actual outcomes on a validation dataset.
- Conduct a detailed analysis of the top three zipcodes identified by the model.
- Assess the model's generalizability and potential for real-world application.

**Deliverables:**
- Evaluation report with insights on model accuracy and reliability.
- Recommendations for the top three zipcodes for investment.



#### 6. Deployment

**Objective:**  
To deploy the predictive model and communicate the results to stakeholders.

**Tasks:**
- Develop a dashboard or interactive tool for stakeholders to explore the results.
- Create documentation and user guides for the deployed solution.
- Present findings and recommendations to potential investors.

**Deliverables:**
- Interactive dashboard or web application.
- Final report and presentation summarizing the project outcomes and recommendations.


# Data Collection/Understanding

Certainly! Here's a step-by-step guide to using Zillow's typical home value datasets for your project:

### Step-by-Step Guide to Using Zillow's Typical Home Value Datasets

#### Step 1: Data Acquisition

**Objective:** Obtain the necessary Zillow home value data and other relevant datasets.

**Tasks:**
1. **Download Zillow Data:**
   - Visit the Zillow Research Data page: [Zillow Data](https://www.zillow.com/research/data/)
   - Look for datasets under "Home Values" such as "ZHVI Single Family Homes Time Series" or similar.
   - Download the relevant CSV files that contain historical home values at the zipcode level for the Baltimore-Washington area.

2. **Obtain Additional Data:**
   - Gather supplementary datasets that might affect home values, such as crime rates, school quality, economic indicators, etc.

#### Step 2: Data Understanding

**Objective:** Explore and understand the structure, content, and initial insights from the datasets.

### HOME VALUES

**Zillow Home Value Index (ZHVI):** A measure of the typical home value and market changes across a given region and housing type. It reflects the typical value for homes in the 35th to 65th percentile range. Available as a smoothed, seasonally adjusted measure and as a raw measure.

Zillow publishes top-tier ZHVI (\$, typical value for homes within the 65th to 95th percentile range for a given region) and bottom-tier ZHVI (\$, typical value for homes within the 5th to 35th percentile range for a given region).

Zillow also publishes ZHVI for all single-family residences (\$, typical value for all single-family homes in a given region), for condo/coops (\$), for all homes with 1, 2, 3, 4 and 5+ bedrooms (\$), and the ZHVI per square foot (\$, typical value of all homes per square foot calculated by taking the estimated home value for each home in a given region and dividing it by the home’s square footage).

**Note:** Starting with the January 2023 data release, and for all subsequent releases, the full ZHVI time series has been upgraded to harness the power of the [neural Zestimate](https://zillow.mediaroom.com/2021-06-15-Zillow-Launches-New-Neural-Zestimate,-Yielding-Major-Accuracy-Gains).

More information about what ZHVI is and how it’s calculated is available on [this overview page](https://www.zillow.com/research/methodology-neural-zhvi-32128). Here’s a handy [ZHVI User Guide](https://www.zillow.com/research/zhvi-user-guide/) for information about properly citing and making calculations with this metric.


![png](images/data-selection.png)

In [1]:
import pandas as pd
import pmdarima as pm
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.tsa.api as tsa

In [3]:
# Load the dat
df_wide = pd.read_csv('data/Zip_zhvi_uc_sfrcondo_tier_0.33_0.67_sm_sa_month.csv')
df_wide

RegionID  SizeRank  ...     2024-05-31     2024-06-30
0         91982         1  ...  508512.211208  508825.477602
1         61148         2  ...  610030.631370  615467.664218
2         91940         3  ...  285741.575496  285701.327874
3         62080         4  ...  458355.229588  459058.239341
4         91733         5  ...  279604.806899  279435.505967
...         ...       ...  ...            ...            ...
26339     63527     39992  ...  230139.106238  235378.350943
26340     64705     39992  ...  102348.053469  105642.477308
26341     78282     39992  ...  193638.478474  194479.590782
26342     67377     39992  ...  372041.654024  375746.672329
26343     80861     39992  ...  117393.470165  119534.346212

[26344 rows x 303 columns]

In [6]:
df_wide['Metro'].value_counts().head(20)

New York-Newark-Jersey City, NY-NJ-PA           830
Chicago-Naperville-Elgin, IL-IN-WI              376
Los Angeles-Long Beach-Anaheim, CA              360
Philadelphia-Camden-Wilmington, PA-NJ-DE-MD     334
Washington-Arlington-Alexandria, DC-VA-MD-WV    314
Pittsburgh, PA                                  296
Boston-Cambridge-Newton, MA-NH                  272
Dallas-Fort Worth-Arlington, TX                 251
Houston-The Woodlands-Sugar Land, TX            219
Minneapolis-St. Paul-Bloomington, MN-WI         217
Detroit-Warren-Dearborn, MI                     213
St. Louis, MO-IL                                211
Atlanta-Sandy Springs-Alpharetta, GA            203
Kansas City, MO-KS                              183
Miami-Fort Lauderdale-Pompano Beach, FL         181
San Francisco-Oakland-Berkeley, CA              166
Seattle-Tacoma-Bellevue, WA                     156
Cincinnati, OH-KY-IN                            149
Phoenix-Mesa-Chandler, AZ                       149
Baltimore-Co

In [8]:
metro_regions = ("Washington-Arlington-Alexandria, DC-VA-MD-WV", "Baltimore-Columbia-Towson, MD")
filter_metro = df_wide['Metro'].isin(metro_regions)
df_wide = df_wide.loc[filter_metro]
df_wide

RegionID  SizeRank  ...     2024-05-31     2024-06-30
78        67182        80  ...  488417.494205  491351.596144
159       67180       165  ...  467623.303339  469785.439440
168       66126       174  ...  659313.373342  654956.248130
193       66705       199  ...  440359.842350  441198.508957
262       66843       268  ...  295369.373645  295506.629210
...         ...       ...  ...            ...            ...
26214     66752     39992  ...  326707.309900  326245.809048
26246     66615     39992  ...  640956.497622  642193.219457
26310     67330     39992  ...  321895.814480  323968.001664
26338     67368     39992  ...  382770.177483  386334.685064
26342     67377     39992  ...  372041.654024  375746.672329

[461 rows x 303 columns]

In [13]:
# Filter for Maryland
df_md = df_wide.loc[df_wide['State'] == 'MD']
df_md['RegionName'] = df_md['RegionName'].astype(str)

# Save the zipcodes
zipcodes = df_md['RegionName'].unique()
len(zipcodes)

/var/folders/rf/vw4r41jd7vd95x1w0dth7v9h0000gp/T/ipykernel_4630/2907951597.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_md['RegionName'] = df_md['RegionName'].astype(str)


281

- There are 281 zipcodes located within MD's Baltimore/DC Metro regions.

## Convert Time Series to Long Form

In [18]:
for i, val in enumerate(df_md.columns):
    print(f"{i}) {val}")

0) RegionID
1) SizeRank
2) RegionName
3) RegionType
4) StateName
5) State
6) City
7) Metro
8) CountyName
9) 2000-01-31
10) 2000-02-29
11) 2000-03-31
12) 2000-04-30
13) 2000-05-31
14) 2000-06-30
15) 2000-07-31
16) 2000-08-31
17) 2000-09-30
18) 2000-10-31
19) 2000-11-30
20) 2000-12-31
21) 2001-01-31
22) 2001-02-28
23) 2001-03-31
24) 2001-04-30
25) 2001-05-31
26) 2001-06-30
27) 2001-07-31
28) 2001-08-31
29) 2001-09-30
30) 2001-10-31
31) 2001-11-30
32) 2001-12-31
33) 2002-01-31
34) 2002-02-28
35) 2002-03-31
36) 2002-04-30
37) 2002-05-31
38) 2002-06-30
39) 2002-07-31
40) 2002-08-31
41) 2002-09-30
42) 2002-10-31
43) 2002-11-30
44) 2002-12-31
45) 2003-01-31
46) 2003-02-28
47) 2003-03-31
48) 2003-04-30
49) 2003-05-31
50) 2003-06-30
51) 2003-07-31
52) 2003-08-31
53) 2003-09-30
54) 2003-10-31
55) 2003-11-30
56) 2003-12-31
57) 2004-01-31
58) 2004-02-29
59) 2004-03-31
60) 2004-04-30
61) 2004-05-31
62) 2004-06-30
63) 2004-07-31
64) 2004-08-31
65) 2004-09-30
66) 2004-10-31
67) 2004-11-30
68) 2004-12

In [24]:
id_vars=['RegionName', 'City', 'State', 'Metro', 'CountyName']
value_cols=  df_md.columns[6:]

df = pd.melt(df_md, id_vars=id_vars , value_vars=value_cols, var_name='Month', value_name='HomeValue')
df

RegionName           City  ...       Month     HomeValue
0          20906  Silver Spring  ...  2000-01-31  1.636624e+05
1          21234      Parkville  ...  2000-01-31  1.148036e+05
2          20878   Gaithersburg  ...  2000-01-31  2.533091e+05
3          20874     Germantown  ...  2000-01-31  1.509697e+05
4          21117   Owings Mills  ...  2000-01-31  1.610868e+05
...          ...            ...  ...         ...           ...
82609      21051          Hydes  ...  2024-06-30  5.278740e+05
82610      20812      Glen Echo  ...  2024-06-30  1.266162e+06
82611      21762   Union Bridge  ...  2024-06-30  3.307898e+05
82612      21052    Fort Howard  ...  2024-06-30  3.262458e+05
82613      20765     Galesville  ...  2024-06-30  6.421932e+05

[82614 rows x 7 columns]

In [25]:
df['CountyName'].value_counts()

Montgomery County        12936
Baltimore County         11172
Anne Arundel County       9996
Prince Georges County     9702
Frederick County          6762
Charles County            6174
Baltimore City            5586
Howard County             5586
Harford County            4704
Calvert County            3822
Queen Annes County        3234
Carroll County            2940
Name: CountyName, dtype: int64

In [30]:
## Process as time series
df['Month'] = pd.to_datetime(df['Month'])#, format='%Y-%m')
df


RegionName           City  ...      Month     HomeValue
0          20906  Silver Spring  ... 2000-01-31  1.636624e+05
1          21234      Parkville  ... 2000-01-31  1.148036e+05
2          20878   Gaithersburg  ... 2000-01-31  2.533091e+05
3          20874     Germantown  ... 2000-01-31  1.509697e+05
4          21117   Owings Mills  ... 2000-01-31  1.610868e+05
...          ...            ...  ...        ...           ...
82609      21051          Hydes  ... 2024-06-30  5.278740e+05
82610      20812      Glen Echo  ... 2024-06-30  1.266162e+06
82611      21762   Union Bridge  ... 2024-06-30  3.307898e+05
82612      21052    Fort Howard  ... 2024-06-30  3.262458e+05
82613      20765     Galesville  ... 2024-06-30  6.421932e+05

[82614 rows x 7 columns]

In [38]:
df_ts = df.set_index(['RegionName', 'Month'])
df_ts = df_ts['HomeValue'].unstack(level=0)#.plot(legend=False, figsize=(12, 6))
df_ts

RegionName          20601          20602  ...          21797          21798
Month                                     ...                              
2000-01-31  171919.785078  139128.334503  ...  266828.885208  170176.720097
2000-02-29  172009.486335  139031.505262  ...  267525.746221  169771.309663
2000-03-31  172342.499785  139225.168544  ...  267588.245616  169500.048425
2000-04-30  172990.719516  139601.597772  ...  268083.564673  169016.255322
2000-05-31  174068.534310  140355.782422  ...  268414.213062  169508.004708
...                   ...            ...  ...            ...            ...
2024-02-29  424642.539119  380632.081252  ...  760177.388112  479539.812379
2024-03-31  427199.776322  382850.984368  ...  764982.099959  483794.634204
2024-04-30  430124.237798  385500.407407  ...  772600.445412  489037.378809
2024-05-31  432242.373849  387428.951724  ...  778871.348058  492842.636674
2024-06-30  433260.478875  388406.738544  ...  781787.773851  494384.365798

[294 rows x 281 columns]

In [39]:
df_ts.columns.name =  None
df_ts

20601          20602  ...          21797          21798
Month                                     ...                              
2000-01-31  171919.785078  139128.334503  ...  266828.885208  170176.720097
2000-02-29  172009.486335  139031.505262  ...  267525.746221  169771.309663
2000-03-31  172342.499785  139225.168544  ...  267588.245616  169500.048425
2000-04-30  172990.719516  139601.597772  ...  268083.564673  169016.255322
2000-05-31  174068.534310  140355.782422  ...  268414.213062  169508.004708
...                   ...            ...  ...            ...            ...
2024-02-29  424642.539119  380632.081252  ...  760177.388112  479539.812379
2024-03-31  427199.776322  382850.984368  ...  764982.099959  483794.634204
2024-04-30  430124.237798  385500.407407  ...  772600.445412  489037.378809
2024-05-31  432242.373849  387428.951724  ...  778871.348058  492842.636674
2024-06-30  433260.478875  388406.738544  ...  781787.773851  494384.365798

[294 rows x 281 columns]

In [41]:
df_ts.index

DatetimeIndex(['2000-01-31', '2000-02-29', '2000-03-31', '2000-04-30',
               '2000-05-31', '2000-06-30', '2000-07-31', '2000-08-31',
               '2000-09-30', '2000-10-31',
               ...
               '2023-09-30', '2023-10-31', '2023-11-30', '2023-12-31',
               '2024-01-31', '2024-02-29', '2024-03-31', '2024-04-30',
               '2024-05-31', '2024-06-30'],
              dtype='datetime64[ns]', name='Month', length=294, freq=None)